libtenzir/builtins/operators/from_to_2.cpp

```diff
@@ -842,13 +840,66 @@ class from_file_impl {
     query_files();
   }
 
+  auto make_behavior() -> from_file_actor::behavior_type {
+    return {
+      [this](atom::get) -> caf::result<table_slice> {
+        return get();
+      },
+      [this](atom::put, table_slice slice) -> caf::result<void> {
+        return put(std::move(slice));
+      },
+      [this](diagnostic diag) {
+        dh_->emit(std::move(diag));
+      },
+      [this](uint64_t nested_operator_index, uuid nested_metrics_id,
+             type schema) {
+        return register_metrics(nested_operator_index, nested_metrics_id,
+                                std::move(schema));
+      },
+      [this](uint64_t nested_operator_index, uuid nested_metrics_id,
+             record metrics) {
+        return handle_metrics(nested_operator_index, nested_metrics_id,
+                              std::move(metrics));
+      },
+      [](const operator_metric& metrics) {
+        // Cannot forward operator metrics from nested pipelines.
+        TENZIR_UNUSED(metrics);
+      },
+    };
+  }
+
+private:
+  auto get() -> caf::result<table_slice> {
+    if (puts_.empty()) {
+      auto rp = self_->make_response_promise<table_slice>();
+      gets_.emplace_back(rp);
+      check_termination();
+      return rp;
+    }
+    auto slice = std::move(puts_.front().first);
+    puts_.front().second.deliver();
+    puts_.pop_front();
+    return slice;
+  }
+
+  auto put(table_slice slice) -> caf::result<void> {
+    if (gets_.empty()) {
+      auto rp = self_->make_response_promise<void>();
+      puts_.emplace_back(std::move(slice), rp);
+      return rp;
+    }
+    gets_.front().deliver(std::move(slice));
+    gets_.pop_front();
+    return {};
+  }
```

Comment:
You might be interested in this piece of code from #5178:

https://github.com/tenzir/tenzir/blob/e780fef24c2fab69bb039b0eb84222cc94f04f13/libtenzir/builtins/operators/window.cpp#L42-L99

Reply:
Definitely makes sense to split this out, also thought about that. Not sure whether I can follow the capacity semantics here, but in any case, I don't think we want to introduce a dependency on this PR here, right? We can get this here merged today.

Reply:
I've written down a note to update this as part of my other PR then.